### ISSUE TITLE
Animation Snapping Only After Export

### ISSUE THREAD

**Issue by StellarGame** [KhronosGroup/glTF-Blender-IO#1901]
I've got a couple of animations on a character + rig, and I'm seeing them snap at the end of their cycles in the threejs and babylonjs viewers, even though the loops are perfect and there's no snapping in blender. The framerate is set to 24fps, the run is 16 frames long and the walk is 29 frames long. My suspicion was originally that it was because they didn't divide evenly into full cycles at 60 fps, but I see the snapping in the viewer even if I set the scene framerate to 30fps for export, so I'm stumped as to what it could be.

Here is the snapping of the animation in Don's GLTF viewer:

https://user-images.githubusercontent.com/130523233/231331478-fcfeb979-a40f-45ff-8506-3fb2d0f02bdb.mp4

These are the animations inside blender. As you can see, there's no snapping during either animation. Does that narrow it down to the exporter and the viewer not agreeing on something?

https://user-images.githubusercontent.com/130523233/231331542-0d492059-065a-458b-ad24-6bca219738fe.mp4

 

These are my exact export settings. The only thing out of view at the top is the format selection.
![Settings](https://cdn.discordapp.com/attachments/1089823371048525904/1092865266729373737/image.png)

On the control rig that I use to make the animations there are keyframes outside the playback range, but when the animation is baked to the game rig it only bakes the keyframes inside the selected range, and I've got that range set to end on the exact right frame for a clean loop, but I'm still seeing the snapping behaviour after exporting from blender 3.5

**Version**
 - OS: Windows
 - Blender Version: 3.5

**Comment by julienduroure** [comment]
Hello,
(I can't check deeply because you didn't provide your file, but here are my assumptions) 

Your animation starts at frame 1, where glTF start from zero. That leads to a very quick snap between zero second and the first animation frame.

Starting Blender 3.6, there is now an option you can use to slide animations during export, to be sure they are starting at zero, and avoid this snap.

![image](https://user-images.githubusercontent.com/1696020/231346411-e66be682-d226-4f4f-969f-1d9c3088ef95.png)

Can you please test with an 3.6 alpha version ?
You can download it here : https://builder.blender.org/download/daily/

**Comment by dizziless** [comment]
Hey Julien, thanks for the suggestion. My animations were indeed beginning on frame 1 like you said, and I was very hopeful -  upon seeing what a bone-headed mistake I made - that a solution was so easily found, but unfortunately I'm still seeing this snapping behaviour. I tried moving the keyframes to frame 0, then baking the animation, pushing it down to the NLA editor, and then exporting, and also baking the animation and then moving the NLA track to frame 0 after being baked, and _then_ exporting, but in both cases, it was still showing. I'm confident that there's nothing wrong with the animation itself, because in blender the loop is flawless, so I'm totally stumped as to where to look for the solution, or even for the problem.

**Comment by scurest** [comment]
This needs a simple .blend that repros the problem.

<details>

<summary>Speculation</summary>

Perhaps it has something to do with this issue.

![Comparison of looping with frame-based vs continuous-time animation](https://user-images.githubusercontent.com/11024420/232176957-e8a59d9e-bed7-4570-a912-8752ce0b3fd1.png)

</details>

**Comment by dizziless** [comment]
I see what you mean. Does this make a difference if the animation is continuous on one rig, and then becomes frame based after being baked to another rig?

**Comment by julienduroure** [comment]
Hello,
Having your .blend file, as already requested multiple times, will help understand exactly the way you perform the animation and how export manage it.
Not really easy to work only with speculations

**Comment by dizziless** [comment]
It turns out that gltf actually _requires_ that the first and last keyframes match to make the animation loop seamlessly, and I guess it just gets rid of the end keyframe somehow? Or, at least, it doesn't repeat and cause a delay like it would in blender, which is kind of counterintuitive to how I would go about creating a seamless loop and rendering it as a video, but there you go. It looks like the problem was exactly as scurest illustrated and that chopping off the last frame so that it wouldn't repeat is unnecessary.

**Comment by StellarGame** [comment]
Yes, it appears this was the main issue. However, we are still generating GLTF errors related to the animation that we're not sure the origin of. Does anyone have an idea where they're coming from?

`<html>
<body>
<p style="color: rgb(0, 0, 0); font-family: &quot;Times New Roman&quot;; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">Report generated by<a href="https://github.com/KhronosGroup/glTF-Validator/">KhronosGroup/glTF-Validator</a><span> </span>2.0.0-dev.3.9.</p>

Error | Message | Pointer
-- | -- | --
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 0 is negative: -0.125. | /animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 0 is negative: -0.125. | /animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 0 is negative: -0.125. | /animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 0 is negative: -0.125. | /animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 0 is negative: -0.125. | /animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 0 is negative: -0.125. | /animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 1 is negative: -0.0833333358168602. | /animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 1 is negative: -0.0833333358168602. | /animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 1 is negative: -0.0833333358168602. | /animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 1 is negative: -0.0833333358168602. | /animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 1 is negative: -0.0833333358168602. | /animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 1 is negative: -0.0833333358168602. | /animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 2 is negative: -0.0416666679084301. | /animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 2 is negative: -0.0416666679084301. | /animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 2 is negative: -0.0416666679084301. | /animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 2 is negative: -0.0416666679084301. | /animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 2 is negative: -0.0416666679084301. | /animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE | Animation input accessor element at index 2 is negative: -0.0416666679084301. | /animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0. | /animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0. | /animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0. | /animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0. | /animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0. | /animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0. | /animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | 1 animation input accessor elements not in ascending order. [AGGREGATED] | /animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | 1 animation input accessor elements not in ascending order. [AGGREGATED] | /animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | 1 animation input accessor elements not in ascending order. [AGGREGATED] | /animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | 1 animation input accessor elements not in ascending order. [AGGREGATED] | /animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | 1 animation input accessor elements not in ascending order. [AGGREGATED] | /animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING | 1 animation input accessor elements not in ascending order. [AGGREGATED] | /animations/8/samplers/2/input

<br class="Apple-interchange-newline">
</body>
</html>Report generated by[KhronosGroup/glTF-Validator](https://github.com/KhronosGroup/glTF-Validator/) 2.0.0-dev.3.9.

Error	Message	Pointer
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 0 is negative: -0.125.	/animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 0 is negative: -0.125.	/animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 0 is negative: -0.125.	/animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 0 is negative: -0.125.	/animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 0 is negative: -0.125.	/animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 0 is negative: -0.125.	/animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 1 is negative: -0.0833333358168602.	/animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 1 is negative: -0.0833333358168602.	/animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 1 is negative: -0.0833333358168602.	/animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 1 is negative: -0.0833333358168602.	/animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 1 is negative: -0.0833333358168602.	/animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 1 is negative: -0.0833333358168602.	/animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 2 is negative: -0.0416666679084301.	/animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 2 is negative: -0.0416666679084301.	/animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 2 is negative: -0.0416666679084301.	/animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 2 is negative: -0.0416666679084301.	/animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 2 is negative: -0.0416666679084301.	/animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NEGATIVE	Animation input accessor element at index 2 is negative: -0.0416666679084301.	/animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0.	/animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0.	/animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0.	/animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0.	/animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0.	/animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	Animation input accessor element at index 3 is less than or equal to previous: 0 <= 0.	/animations/8/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	1 animation input accessor elements not in ascending order. [AGGREGATED]	/animations/7/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	1 animation input accessor elements not in ascending order. [AGGREGATED]	/animations/7/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	1 animation input accessor elements not in ascending order. [AGGREGATED]	/animations/7/samplers/2/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	1 animation input accessor elements not in ascending order. [AGGREGATED]	/animations/8/samplers/0/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	1 animation input accessor elements not in ascending order. [AGGREGATED]	/animations/8/samplers/1/input
ACCESSOR_ANIMATION_INPUT_NON_INCREASING	1 animation input accessor elements not in ascending order. [AGGREGATED]	/animations/8/samplers/2/input`

**Comment by julienduroure** [comment]
Closing this ticket, as needed file to investigate is not provided since more than a month.
Feel free to open a new ticket with all required information needed.
